Check that signatures of @usableFromInline declarations only reference @usableFromInline or public types

slavapestov opened the pull request:
We were performing these checks for references from the body of inlinable functions as part of availability checking, but we also have to do similar checks in access level checking for types that are referenced from the signature of the function itself.

Also, allow type aliases to be @usableFromInline, and ensure that only @usableFromInline type aliases are referenced from inlinable bodies.

This plugs some holes in the model that could cause linker errors on invalid code. Fixes rdar://problem/32722385, rdar://problem/39532028,

slavapestov committed:
Sema: Declaration signatures should respect @usableFromInline

Previously we made sure that inlinable function bodies only
reference @usableFromInline or public declarations, but we
also have to make sure that @usableFromInline declarations
only reference other declarations that are at least as
visible.

Otherwise, you could, for example, define a @usableFromInline
typealias which referenced an internal type, and then
reference the typealias from a @usableFromInline function body.

slavapestov committed:
AST: Some tweaks for ValueDecl::isUsableFromInline()

Make this method public and fix it to do the right thing for
associated types.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1317,20 +1328,25 @@ ERROR(unsupported_nested_protocol,none,
 ERROR(type_alias_underlying_type_access,none,
       "type alias %select{must be declared %select{"
-      "%select{private|fileprivate|internal|%error|%error}2|private or fileprivate}3"
+      "%select{private|fileprivate|internal|%error|%error}1|private or fileprivate}3"
```
(here and below) This was correct the way it was before. Argument 1 is the current access; argument 2 is the required access.

slavapestov replied:
Are you sure? When isExplicit is true, %1 is the decl's getFormalAccess(), when isExplicit is false, %1 is the AccessScope's `requiredAccessForDiagnostics()`. On the other hand, %2 is the AccessScope's `accessLevelForDiagnostics()`.

I noticed some inconsistency here and cleaned it up. Some diagnostics were using %2 here for the isExplicit == false case, but this appeared to generate a wrong diagnostic because of the distinction between `requiredAccessForDiagnostics()` and `accessLevelForDiagnostics()` with private vs fileprivate. This change and tests for it was in a separate patch so it should be easier to review that way.

jrose-apple replied:
Note the difference in messages: "must be declared X" vs. "cannot be declared Y". There might be mistakes at some of the use sites, but they're definitely not supposed to be the same value.

slavapestov replied:
"When isExplicit is true, %1 is the decl's getFormalAccess(), when isExplicit is false, %1 is the AccessScope's requiredAccessForDiagnostics()" -- so they're different values :)

So using %1 vs %2 in the !isExplicit case is almost equivalent, except for private vs fileprivate.

jrose-apple replied:
I see: you changed the call site too. Please don't do this; it muddles the notion of what each parameter is used for. At that point I'd rather just use different diagnostics.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1316,16 +1316,21 @@ namespace {
   Optional<AccessScope> Scope = AccessScope::getPublic();
 
   AccessScopeChecker(const DeclContext *useDC,
+                     bool treatUsableFromInlineAsPublic,
```
You're definitely starting to convince me that UsableFromInline is really a new access level.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1842,11 +1855,21 @@ static void highlightOffendingType(TypeChecker &TC, InFlightDiagnostic &diag,
 
+  if (checkUsableFromInline) {
+    if (auto *VD = dyn_cast<ValueDecl>(owner)) {
+      if (VD->getFormalAccess() != AccessLevel::Internal)
```
This is a suspicious condition—this could be above `internal` or it could be below `internal`.

slavapestov replied:
`checkUsableFromInline` means *only* check @usableFromInline and nothing else. So we call this function once with `checkUsableFromInline=false` if the decl is not @usableFromInline; if it is, we call it twice, first with `false` then with `true`.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1911,21 +1942,20 @@ static void checkGenericParamAccess(TypeChecker &TC,
-        !isa<ModuleDecl>(accessScope.getDeclContext()) &&
-        TC.getLangOpts().isSwiftVersion3()) {
-      downgradeToWarning = DowngradeToWarning::Yes;
-    }
```
What happened to this code?

slavapestov replied:
This was a separate patch -- it was duplicated twice so I moved it.

jrose-apple replied:
…and now it's applied to more places. Maybe it doesn't matter, though.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1911,21 +1942,20 @@ static void checkGenericParamAccess(TypeChecker &TC,
   bool isExplicit =
     owner->getAttrs().hasAttribute<AccessControlAttr>() ||
-    owner->getDeclContext()->getAsProtocolOrProtocolExtensionContext();
+    isa<ProtocolDecl>(owner->getDeclContext());
```
Oops, good catch! Is there a test for this?

slavapestov replied:
Yep, see https://github.com/apple/swift/pull/16586/files#diff-8ace5fe1c031aa385557a1c025b63437

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1939,13 +1969,16 @@ static void checkGenericParamAccess(TypeChecker &TC,
-/// Checks the given declaration's access to make sure it is valid given the way
-/// it is defined.
+/// Checks the given declaration's signature does not reference any other
+/// declarations that are less visible than the declaration itself.
```
In the access control code I've tried to be good about distinguishing "is visible" from "has access to". Something declared `public` in a module that hasn't been (transitively) imported is not visible, but the client code still has access to it.

jrose-apple replied:
(this one's still outstanding too)

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1994,13 +2027,14 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
           bool isExplicit =
-            theVar->getAttrs().hasAttribute<AccessControlAttr>();
+            theVar->getAttrs().hasAttribute<AccessControlAttr>() ||
+            isa<ProtocolDecl>(theVar);
```
Definitely needs a new test if existing ones don't cover it.

slavapestov replied:
See https://github.com/apple/swift/pull/16586/files#diff-8ace5fe1c031aa385557a1c025b63437, I think I covered it

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2175,14 +2223,16 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
     auto SD = cast<StructDecl>(D);
-    checkGenericParamAccess(TC, SD->getGenericParams(), SD);
+    checkGenericParamAccess(TC, SD->getGenericParams(), SD,
+                            /*checkUsableFromInline=*/false);
```
(here and elsewhere) Why is this `false` and not inherited? Can you leave a comment to that effect?

slavapestov replied:
That's correct. `checkAccessControl()` only checks access control rules. The separate `checkUsableFromInline()` performs the same visitation of all TypeLocs, calling checkTypeAccess() with `checkUsableFromInline=true`.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+      if (anyVar->getFormalAccess() != AccessLevel::Internal)
+        return;
+      if (!anyVar->isUsableFromInline())
+        return;
```
This "internal and usable from inline" check shows up a lot. Maybe worth a helper function, even though it's so simple?

slavapestov replied:
Yeah. I'm going to drop the assertion from isUsableFromInline() that the access level is internal, and just return false if its not.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+    };
+
+    std::for_each(assocType->getInherited().begin(),
+                  assocType->getInherited().end(),
```
This needs to check requirements, not just the inheritance clause, right? (Also, `llvm::for_each`.)

slavapestov replied:
That's a good point. As you correctly noted this is copy and pasted from checkAccessControl() so probably the existing code was wrong too.

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2437,6 +2498,368 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
+      Type rawType = ED->getRawType();
+      auto rawTypeLocIter = std::find_if(ED->getInherited().begin(),
+                                         ED->getInherited().end(),
+                                         [&](TypeLoc inherited) {
```
Okay this is really starting to seem like too much code duplicated from the regular access checking.

slavapestov replied:
Yeah. Perhaps we can factor out a visitor that walks the TypeLocs in a Decl?

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4508,6 +4934,7 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
 
     TC.checkDeclAttributes(SD);
     checkAccessControl(TC, SD);
+    checkUsableFromInline(TC, SD);
```
Do we need to do both? Can't `checkAccessControl` handle it?

slavapestov replied:
Yes, I could merge the two into one function. It would eliminate the duplicated logic for visting the TypeLocs, at the cost of making the whole function longer. However, we could then split it up into smaller functions for each Decl kind or something, maybe even turn it into a visitor. Sounds good?

jrose-apple reviewed (dismissed):
Whew, that was more than I thought it would be! I do have some concerns about this implementation, though.

slavapestov commented:
Source compat failure:

```
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/Chatto/ChattoAdditions/Source/Chat Items/PhotoMessages/Views/PhotoMessageCollectionViewCellDefaultStyle.swift:92:36: error: type alias 'Class' is internal and cannot be referenced from a default argument value
        bubbleMasks: BubbleMasks = Class.createDefaultBubbleMasks(),
                                   ^
/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/Chatto/ChattoAdditions/Source/Chat Items/PhotoMessages/Views/PhotoMessageCollectionViewCellDefaultStyle.swift:28:15: note: type alias 'Class' is not public
    typealias Class = PhotoMessageCollectionViewCellDefaultStyle
              ^
```

Looks like I need to disable the new checks in `-swift-version < 4.2`.

jrose-apple commented:
Less than 5, please. We don't currently have any language changes that are gated on 4.2 besides SDK changes, right?

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2419,7 +2410,7 @@ static void checkAccessControl(TypeChecker &TC, const Decl *D) {
       return;
     for (auto &P : *EED->getParameterList()) {
-      checkTypeAccess(TC, P->getTypeLoc(), P,
+      checkTypeAccess(TC, P->getTypeLoc(), EED->getParentEnum(),
```
This one seems wrong; it should just be `EED`, right? (Because some day we might have enum cases less visible than the enum.)

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -2082,7 +2082,8 @@ SourceLoc ValueDecl::getAttributeInsertionLoc(bool forModifier) const {
 bool ValueDecl::isUsableFromInline() const {
-  assert(getFormalAccess() == AccessLevel::Internal);
+  if (getFormalAccess() != AccessLevel::Internal)
+    return false;
```
Unhappy with this. If you're going to make this an API that can be called everywhere, shouldn't it return true for `open` and `public` decls as well?

jrose-apple reviewed (commented):
Comments for everything except the last Sema change.

slavapestov committed:
Sema: Move access control checking to TypeCheckAccess.cpp

slavapestov committed:
Sema: Introduce AccessControlChecker

slavapestov committed:
Sema: Rework AccessControlChecker into a class hierarchy

AccessControlCheckerBase
|- AccessControlChecker
|- UsableFromInlineChecker

This factors out the checkUsableFromInline parameter that
is passed around all over the place.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
This has been open for far too long. Since the latest iteration moves a lot of code to a new source file, I'm going to merge it now to avoid more merge conflicts. I'm not quite done extracting out the common visitor, but if you squint you will be able to guess where it's going to go.
